### PR TITLE
chore(ui): Prepare for react router v6 upgrade

### PR DESCRIPTION
**Pull request by bradr5**
### Description

Satisfies step 2 in the [official react router v6 upgrade documentation](https://reactrouter.com/en/main/upgrading/v5#upgrade-to-react-router-v51)

1. ~~Upgrade to React v16.8 or greater~~
2. **Upgrade to React Router v5.1**
    * **Remove `<Redirect>`s inside `<Switch>`**
    * **Refactor custom `<Route>`s**
3. Upgrade to React Router v6

More specifically it satisfies the following steps:

- Use `<Route children>` instead of `<Route render>` and/or `<Route component>` props
- Use hooks API to access router state like the current location and params
- Replace all uses of `withRouter` with hooks
- Remove `<Redirect>`s inside of `<Switch>`

### User-facing documentation

- [ ] CHANGELOG update is not needed
- [ ] documentation PR is not needed

#### How I validated my change

Verify links and routes still work as intended

### CODE REVIEW

File: ui/apps/platform/src/Containers/ComplianceEnhanced/Schedules/ScanConfigsPage.tsx [hunk 1]
```diff
@@ -25,34 +25,31 @@ function ScanConfigsPage() {
...
+                        return <Redirect to={complianceEnhancedSchedulesPath} />;
+                    }}
+                />
+            </Route>
+            <Route exact path={scanConfigDetailsPath}>
+                {() => (
+                    <ScanConfigDetailPage
+                        hasWriteAccessForCompliance={hasWriteAccessForCompliance}
+                    />
+                )}
```
Comment (dvail) [inline ui/apps/platform/src/Containers/ComplianceEnhanced/Schedules/ScanConfigsPage.tsx:51]: This seems a little odd, providing an anonymous function as a child element instead of the JSX directly. Why is this needed versus rendering `<ScanConfigDetailPage>` as a direct child?

File: ui/apps/platform/src/Containers/ConfigManagement/SidePanel/BreadCrumbs.js [hunk 2]
```diff
@@ -85,9 +84,12 @@ const getMaxWidthClass = (length) => {
...
 
 const BreadCrumbLinks = (props) => {
+    const location = useLocation();
+    const match = useRouteMatch();
+
     // disable because unused history might be specified for rest spread idiom.
     /* eslint-disable no-unused-vars */
-    const { className, match, location, history, ...params } = props;
+    const { className, ...params } = props;
     /* eslint-enable no-unused-vars */
```
Comment (dvail) [inline ui/apps/platform/src/Containers/ConfigManagement/SidePanel/BreadCrumbs.js:93]: Nit: can we remove these eslint suppression comments now?

File: ui/apps/platform/src/Containers/AppPage.tsx [hunk 3]
```diff
@@ -20,14 +20,21 @@ function AppPage(): ReactElement {
             <AppPageTitle />
             <AppPageFavicon />
             <Switch>
-                <Route path={loginPath} component={LoginPage} />
-                <Route
-                    path={authorizeRoxctlPath}
-                    render={(props) => <LoginPage {...props} authorizeRoxctlMode />}
```
Comment (dvail) [inline ui/apps/platform/src/Containers/AppPage.tsx:26]: For my own understanding - the `...props` here are just the match/history/location props provided by react-router right? We don't lose anything by not forwarding these props in the change?

File: ui/apps/platform/src/Containers/AppPage.tsx [hunk 4]
```diff
@@ -20,14 +20,21 @@ function AppPage(): ReactElement {
...
-                />
-                <Route path={testLoginResultsPath} component={TestLoginResultsPage} />
-                <Route path={authResponsePrefix} component={LoadingSection} />
-                <Route component={AuthenticatedRoutes} />
+                <Route path={loginPath}>
+                    <LoginPage authProviders={undefined} />
+                </Route>
+                <Route path={authorizeRoxctlPath}>
+                    <LoginPage authorizeRoxctlMode authProviders={undefined} />
+                </Route>
```
Comment (dvail) [inline ui/apps/platform/src/Containers/AppPage.tsx:28]: Not a change request, but what are your thoughts on the following alternatives?

- Removing `isRequired` from the `authProviders` prop in the prop types
- Passing `[]` as the initial value instead of `undefined`. (This might be a change in runtime behavior, but conforms with the expected type that is used within the `LoginPage` component)

**Review by dvail — COMMENTED** [review]
Overall this looks good, thanks for powering through the 86 file change! 😄 

One request to avoid a bad conflict with the ongoing yarn => npm conversion, but the rest of the comments are nits and minor questions.

There are some e2e failures that indicate something isn't quite right, but I think once those are fixed and we do a small team bug-bash this should be close to merge-able.

File: ui/apps/platform/package.json [hunk 5]
```diff
@@ -71,7 +71,6 @@
         "react-responsive": "^9.0.2",
         "react-router-dom": "^5.3.4",
         "react-router-hash-link": "^2.4.3",
-        "react-router-prop-types": "^1.0.5",
```
[the remark this answers is not in the document]
Comment (bradr5) [reply ui/apps/platform/package.json:74]: Ah, good call out. That makes sense to me.

File: ui/apps/platform/src/Components/PatternFly/ErrorBoundary/ErrorBoundary.tsx [hunk 6]
```diff
@@ -23,7 +23,7 @@ type State =
           errorInfo: ErrorInfo;
       };
 
-class ErrorBoundary extends Component<Props, State> {
+class ErrorBoundaryClass extends Component<Props, State> {
```
[the remark this answers is not in the document]
Comment (bradr5) [reply ui/apps/platform/src/Components/PatternFly/ErrorBoundary/ErrorBoundary.tsx:26]: I actually original had the naming you described but switched it because I wasn't sure if we typically name a file differently from the default export. I went back and forth on this though, so I'm glad to get a second opinion. I'll revert it.

File: ui/apps/platform/src/Containers/AppPage.tsx (hunk 3 above)
Comment (bradr5) [reply ui/apps/platform/src/Containers/AppPage.tsx:26]: I _believe_ so. The route props are also passed regardless here but I'm guessing the original author wanted to be more explicit. Although neither location, match, or history is being used in the `LoginPage` component, instead `window.location` is being used :thinking:

Here's at least the link to [render props documentation](https://v5.reactrouter.com/web/api/Route/render-func)

Comment (pedrottimark) [reply ui/apps/platform/src/Containers/AppPage.tsx:26]: @bradr5 did TypeScript require `authProviders={undefined}` prop here?

Does explicit prop conflict with access to Redux store here:
https://github.com/stackrox/stackrox/blob/master/ui/apps/platform/src/Containers/Login/LoginPage.js#L309

Comment (bradr5) [reply ui/apps/platform/src/Containers/AppPage.tsx:26]: TypeScript was requiring `authProviders` be set here since it's required in `LoadingOrForm` inside of `LoginPage.js`. However, regardless of whats being passed as a prop, the redux store `selectors.getLoginAuthProviders` is overriding whatever prop is being passed. Before the changes I made, it appears that `undefined` was being passed but typescript wasn't catching that as an issue since we were using the `Route` -> `component` prop. That's one of the benefits of making the incremental change before fully upgrading to v6.

It does seem like @dvail suggestion to default auth providers to an empty array makes the most sense. `undefined` would actually break the component if it wern't for redux and `mapStateToProps`.

File: ui/apps/platform/src/Containers/AppPage.tsx (hunk 4 above)
Comment (bradr5) [reply ui/apps/platform/src/Containers/AppPage.tsx]: I like option 2 based off what I wrote here https://github.com/stackrox/stackrox/pull/12183#discussion_r1765539168

File: ui/apps/platform/src/Containers/AppPage.tsx (hunk 3 above)
Comment (pedrottimark) [reply ui/apps/platform/src/Containers/AppPage.tsx:26]: Yes, super. I agree with you and David. Thank you for investigating.

**Review by pedrottimark — APPROVED** [review]
Amazing effort, Brad <3
